### PR TITLE
validation library

### PR DESCRIPTION
**Pull request by estroz**
Moved from [dweepgogia/new-manifest-verification](https://github.com/dweepgogia/new-manifest-verification).

Implementation of the library described in #2.

Notes:

* See my [comment](https://github.com/operator-framework/api/pull/1#issuecomment-534638383) about `api` repo structure, which is not followed here but could/should be.
* Quite a bit more validation logic needs to be ported from operator-courier.

/cc @gallettilance @shawn-hurley @ecordell

### CODE REVIEW

File: pkg/validation/interfaces/validator.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+// ValidationFuncs is a set of validation functions.
+type ValidatorFuncs []ValidatorFunc
+
+// Validator is an interface for validating arbitrary objects. A Validator
+// returns a set of functions that each validate some underlying object.
+// This allows the implementer to easily break down each validation step
+// into discrete units if they wish, and perform either lazy or active
+// validation.
+type Validator interface {
```
Comment (njhale) [inline pkg/validation/interfaces/validator.go]: Some thoughts on the structure here. Could `Validator` itself be the thing that does a validation?

```golang
type Validator interface {
   Validate(objs ...interface{}) (results ...errors.ManifestResult)
}
```

Then `ValidatorFunc` could be a function type that implements `Validator`.

```golang
type ValidatorFunc func(objs ...interface{}) (...errors.ManifestResult)

func (v ValidatorFunc) Validate(objs ...interface{}) (results ...errors.ManifestResult) {
    return v(objs...)
} 
```

`Validators` can still be a list of validators, but can also implement `Validate` and aggregate the result.

```golang
type Validators []Validator

func (v Validators) Validate(objs ...interface{}) (results ...errors.ManifestResult) {
    for validate := range v { 
        results = append(results, validate(objs...)...)
    }

    return
}

```

File: pkg/validation/internal/bundle.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
+package internal
```
Comment (njhale) [inline pkg/validation/internal/bundle.go]: Is there a reason we want to [prevent others from using this code](https://golang.org/doc/go1.4#internalpackages)?

File: pkg/validation/interfaces/validator.go (hunk 1 above)
Comment (estroz) [reply pkg/validation/interfaces/validator.go]: That's what I was originally thinking because it's simple. The interface usage isn't _that_ much different either:

```go
type customValidator struct{}

func (v customValidator) GetFuncs(objs ...interface{}) (funcs interfaces.ValidatorFuncs) {
	for _, obj := range objs {
		...
	}
	return funcs
}

func (v customValidator) Validate(objs ...interface{}) (results []errors.ManifestResult) {
	for _, obj := range objs {
		...
	}
	return results
}

func tester() {
	objs := []interface{}{...}
	// Validator with GetFuncs
	results := DefaultValidators(customValidator{}).Apply(objs)

	// Validator with Validate
	validators := append(DefaultValidators(), customValidator{})
	results := validators.Validate(objs)
}
```

However, the `GetFuncs()` method is more parallelizable, and we'll get a better speedup assuming the number of validators N << number of objects M, and that validation is a resource-intensive tast. That being said, I don't think the speedup will be large because N and M won't be more than one order of magnitude different in most cases. And it won't matter at all if objects are being validated actively (not lazily).

In the end I'm fine with either interface layout, since they're not too different and provide the same features.

File: pkg/validation/interfaces/validator.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+// Validators is a set of Validator's that can be run via Apply.
+type Validators []Validator
+
+// Apply collects validator functions from each Validator in vals by collecting
+// the appropriate functions for each obj in objs, then invokes them,
+// collecting the results. Use Apply in code that:
+// - Uses more than one Validator in one call.
+// - Want active validation.
+func (vals Validators) Apply(objs ...interface{}) (results []errors.ManifestResult) {
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/interfaces/validator.go]: No validator should modify an object, so I would say so. We could enforce that objects are not pointers. If a `Validator`'s underlying type stores some information that may be accessed concurrently, then there's a problem. We should note that `Validator`s must be threadsafe in the method comments.

However the `GetFuncs()` concept promotes closures, which are likely not sharing non-threadsafe data. The default `Validator` implementations are good examples of this.

File: pkg/validation/interfaces/validator.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	queue := make(chan []errors.ManifestResult, 1)
+	wg := &sync.WaitGroup{}
+	wg.Add(len(vals))
+	for _, val := range vals {
+		go func(v Validator) {
+			funcResults := []errors.ManifestResult{}
+			for _, result := range v.GetFuncs(objs...).runP() {
+				if result.HasError() || result.HasWarn() {
+					funcResults = append(funcResults, result)
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/interfaces/validator.go]: We don't know how many funcs are returned by `v.GetFuncs()` so we can't add the correct delta outside of the goroutine, i.e. we can't know the correct number of times to call `wg.Done()`.

File: cmd/operator-verify/main.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	"os"
+
+	manifests "github.com/operator-framework/api/cmd/operator-verify/manifests"
+
+	"github.com/spf13/cobra"
+)
+
+func main() {
+	rootCmd := &cobra.Command{
+		Use:   "operator-verify",
```
[the remark this answers is not in the document]
Comment (estroz) [reply cmd/operator-verify/main.go]: I'm fine with that, `operator-verify` was its name when I inherited the project.

@gallettilance @shawn-hurley @awgreene thoughts on naming?

File: README.md [hunk 6]
```diff
@@ -1,2 +1,18 @@
 # api
-Contains the API definitions used by OLM and Marketplace
+
+Contains the API definitions used by [Operator Lifecycle Manager][olm] (OLM) and Marketplace
+
+## `pkg/validation`: Operator Manifest Validation
+
+`pkg/validation` exposes a convenient set of interfaces to validate Kubernetes object manifests, primarily for use in an Operator project.
+
+Additionally this package defines a valid Operator [manifests format][registry-manifests] by exposing a set of `interfaces.Validator`'s to verify them and their constituent manifests. These bundles typically consist of a [Package Manifest][pkg-man] and a set of versioned [bundles][registry-manifests]. Each bundle contains a [ClusterServiceVersion][csv] and one or more [CustomResourceDefinition][crd]'s.
```
[the remark this answers is not in the document]
Comment (estroz) [reply README.md]: Thanks for the tip. I want this library to be general enough that modification will not be extensive when a different bundle/manifest format is implemented.

File: pkg/validation/internal/package_manifest.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		return errs
+	}
+	if pkg.DefaultChannelName == "" {
+		errs = append(errs, errors.WarnInvalidPackageManifest("default channel not found", pkg.PackageName))
+	}
+
+	seen := map[string]struct{}{}
+	for i, c := range pkg.Channels {
+		if c.Name == "" {
+			errs = append(errs, errors.ErrInvalidPackageManifest(fmt.Sprintf("channel %d name is empty", i), pkg.PackageName))
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/package_manifest.go]: Yes, here I assume there's no other identifying information for a channel so channel number is the only piece of information available to identify which channel is invalid.

File: pkg/validation/validation.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
+package validation
+
+import (
+	interfaces "github.com/operator-framework/api/pkg/validation/interfaces"
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/validation.go]: My text editor can't find the proper package without them for certain package names :smile:. I'll remove these before pushing.

File: cmd/operator-verify/main.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	manifests "github.com/operator-framework/api/cmd/operator-verify/manifests"
+
```
[the remark this answers is not in the document]
Comment (estroz) [reply cmd/operator-verify/main.go]: No, but I prefer to separate intra-repo from extra-repo imports. I can merge them all if you prefer.

File: cmd/operator-verify/main.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+)
+
+func main() {
+	rootCmd := &cobra.Command{
+		Use:   "operator-verify",
+		Short: "Operator manifest validation tool",
+		Long: `operator-verify is a CLI tool for the Operator manifest validation
+	library, which provides functions to validate operator manifest bundles `,
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			fmt.Println("Initializing verification CLI tool...")
```
[the remark this answers is not in the document]
Comment (estroz) [reply cmd/operator-verify/main.go]: Nope, removing.

File: pkg/validation/internal/test_suite.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+			if numErrors != c.numErrs || numWarns != c.numWarns {
+				t.Errorf("%s: expected %d errors and %d warnings, got: %d and %d",
+					c.description, c.numErrs, c.numWarns, numErrors, numWarns)
+			}
+			allErrors := append(result.Errors, result.Warnings...)
+			if !deepEqualErrors(c.errors, allErrors) {
+				t.Errorf("%s:\nexpected:\n\t%#v\ngot:\n\t%#v", c.description, c.errors, allErrors)
+			}
+		}
+	} else if result.HasError() || result.HasWarn() {
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/test_suite.go]: Added a `c.wantWarn` field.

File: pkg/validation/internal/bundle.go (hunk 2 above)
Comment (estroz) [reply pkg/validation/internal/bundle.go]: They can use this code through exported `pkg/validation/validation.go` functions. We can make sub-validation functions in `pkg/validation/internal` into `Validator`s themselves and export them too.

File: cmd/operator-verify/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (estroz) [reply cmd/operator-verify/main.go]: I'm actually going to remove this since this repo probably shouldn't be a source of CLI tools. Rather, `operator-sdk` should handle this functionality. At the very least, there will be something like mentioned in the [bundle EP](https://github.com/openshift/enhancements/blob/dec2726b852d7fadd3c7e0419edb24e25ecd67d7/olm/operator-bundle.md#validate-an-operator-bundle).

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (njhale) [reply README.md]: So, if a new format comes out, is the intended workflow to build a new validator implementation to support it?

File: cmd/operator-verify/main.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (njhale) [reply cmd/operator-verify/main.go]: In OLM we use `goimports` which actually does something similar, just with the position of the intra/extra imports flipped from what you have here. Do we want to make it the org wide formatter?

File: cmd/operator-verify/manifests/cmd.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+func NewCmd() *cobra.Command {
+	return &cobra.Command{
+		Use:   "manifests",
+		Short: "Validate manifests directory",
+		Long: `Validates all manifests in a directory that have validators in the
```
[the remark this answers is not in the document]
Comment (njhale) [reply cmd/operator-verify/manifests/cmd.go]: `\`` declares a multiline string literal, which lets write more natural line breaks. I would prefer this to writing `\\n` everywhere.

File: go.mod [hunk 13]
```diff
@@ -0,0 +1,43 @@
+module github.com/operator-framework/api
+
+go 1.12
+
+replace (
+	// Pin openshift version to 4.2 (uses kube 1.14)
+	github.com/openshift/api => github.com/openshift/api v3.9.1-0.20190717200738-0390d1e77d64+incompatible
+	github.com/openshift/client-go => github.com/openshift/client-go v0.0.0-20190627172412-c44a8b61b9f4
+
+	// Pin kube version to 1.14
```
Comment (njhale) [inline go.mod]: We can bump to kube 1.16 now since OLM has already moved forward.

File: pkg/validation/interfaces/validator.go (hunk 1 above)
Comment (njhale) [reply pkg/validation/interfaces/validator.go]: > However, the GetFuncs() method is more parallelizable

Wouldn't they be equivalently parallelizable? You could define the `Validators.Validate` method in such a way that it parallelizes calls to its validators. Alternatively, you could define a new `ParallelValidator` that has `Validators` as a field and fires its elements off in parallel instead of calling its `Validate` method.

File: pkg/validation/interfaces/validator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/validation/interfaces/validator.go]: > We don't know how many funcs are returned by v.GetFuncs() so we can't add the correct delta outside of the goroutine

I'm going to assume that most of the work goes on when executing a validation function, rather than in getting them with `v.GetFuncs()` (correct me if I'm wrong). Using this assumption, we can synchronously get the functions, and execute them concurrently. 

I decided to add some more detail and ended up making something simpler than what I originally described. If we assume _no changes_ to the current interfaces:

```go
func (vals Validators) Apply(objs ...interface{}) (results []errors.ManifestResult) {
	if len(vals) == 0 {
		return nil
	}

        var wg sync.WaitGroup
        var mux sync.Mutex
	for _, val := range vals {
                // Assume it doesn't take a long time to get the set of functions
		for _, fn := range v.GetFuncs(objs...) {
                        // Synchronously add to the WaitGroup for each function we need to evaluate
                        wg.Add(1)

                        // Spin up a goroutine to handle the work and continue to the next function
                        go func() {
                             // Ensure we let the outside world know when we finish
                             defer wg.Done()

                             // Do the long-running work
                             result := fn()
			    if result == nil || !result.HasError() && !result.HasWarn() {
                                 return
			    }

                             // Aggregate the results synchronously
                             mux.Lock()
                             defer mux.Unlock()
                             results = append(results, result)

                             return
			}
		}
	}
        
        // Wait for all goroutines to return
	wg.Wait()

        // at this point, results is populated with all validator results
	return
}
```

Let me know if this makes sense, or maybe I'm missing something subtle.

File: pkg/validation/interfaces/validator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/validation/interfaces/validator.go]: These are fair points. My question was more in the spirit of being cautious of premature optimization.

**Comment by njhale** [comment]
@kevinrizza @ecordell, did we decide that we're abandoning this repo. If so we should make plans to move this PR into its own validation repo or OLM proper.

(hopefully this question doesn't rekindle the entire conversation...)

**Comment by kevinrizza** [comment]
> @kevinrizza @ecordell, did we decide if we're abandoning this repo? If so, we should make plans to move this PR into its own validation repo or OLM proper.
> 
> (hopefully this question doesn't rekindle the entire conversation...)

There is an active effort to create an API version for the entire operator-framework. I think once that conversation gets further along we may end up wanting to merge this (either in the current state or otherwise) so for now maybe let's just leave it open and reevaluate later?

**Comment by njhale** [comment]
> There is an active effort to create an API version for the entire operator-framework. I think once that conversation gets further along we may end up wanting to merge this (either in the current state or otherwise) so for now maybe let's just leave it open and reevaluate later?

Sure, we can leave this here for now -- my comment was spurred by the decision to place net-new OLM APIs directly in its repository (instead of here).

File: pkg/validation/interfaces/validator.go (hunk 1 above)
Comment (estroz) [reply pkg/validation/interfaces/validator.go]: Going with the simpler, non-parallelized approach. We can re-assess later if we feel improvements are needed.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (estroz) [reply README.md]: We would potentially re-write the bundle validator, or write some pull/unpack logic and expose it to users.

File: pkg/internal/bundle.go [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+		pkg:     l.pkg,
+		bundles: l.bundles,
+	}, nil
+}
+
+func (l manifests) GetPackageManifest() *registry.PackageManifest {
+	return l.pkg
+}
+
+func (l manifests) GetBundles() (bundles []*registry.Bundle) {
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/internal/bundle.go]: operator-registry uses [`*Bundle`](https://github.com/operator-framework/operator-registry/blob/master/pkg/sqlite/directory.go#L124), a convention I stuck with here. Also this is an internal package so I figured pointer vs. value wouldn't matter much. Should I make this a slice of values?

File: pkg/internal/bundle.go [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+	for _, bundle := range l.bundles {
+		bundles = append(bundles, bundle)
+	}
+	return bundles
+}
+
+func (l manifests) GetBundleForVersion(version string) (*registry.Bundle, error) {
+	if _, err := semver.Parse(version); err != nil {
+		return nil, errors.Wrapf(err, "error getting bundle for version %q", version)
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/internal/bundle.go]: Any error from [`Parse()`](https://github.com/blang/semver/blob/master/semver.go#L264). This invocation is done to validate the version string. We could relax this check if we do not always expect a bundle to be versioned with semver. The way I'm indexing CSV's in this parser is by CSV `spec.version`.

File: pkg/validation/validation.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/validation.go]: Apparently the compiler complains that `pkg/internal` is not defined without the alias.

```console
$ go build -o operator-verify ./cmd/operator-verify/main.go 
# github.com/operator-framework/api/pkg/manifests
pkg/manifests/directory.go:6:2: imported and not used: "github.com/operator-framework/api/pkg/internal" as manifests
pkg/manifests/directory.go:17:20: undefined: internal
```

File: go.mod [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+	k8s.io/kubernetes => k8s.io/kubernetes v1.14.5-beta.0.0.20190708100021-7936da50c68f
+	sigs.k8s.io/structured-merge-diff => sigs.k8s.io/structured-merge-diff v0.0.0-20190302045857-e85c7b244fd2
+)
+
+require (
+	github.com/blang/semver v3.5.1+incompatible
+	github.com/ghodss/yaml v1.0.0
+	github.com/hashicorp/golang-lru v0.5.3 // indirect
+	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
+	github.com/operator-framework/operator-lifecycle-manager v0.0.0-20190725173916-b56e63a643cc
```
[the remark this answers is not in the document]
Comment (estroz) [reply go.mod]: I would have to duplicate the ClusterServiceVersion and related types here to remove this dependency. Are you ok with this?

We may have to do the same thing for operator-registry (`registry.Bundle` and some parsing types).

File: go.mod (hunk 13 above)
Comment (estroz) [reply go.mod]: Can we do this in a follow-up PR?

File: go.mod [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+	gopkg.in/inf.v0 v0.9.1 // indirect
+	k8s.io/apiextensions-apiserver v0.0.0-20181204003618-e419c5771cdc
+	k8s.io/apimachinery v0.0.0-20190717022731-0bb8574e0887
+	k8s.io/client-go v8.0.0+incompatible
+)
+
+// TODO(estroz): change this to point to the main repo once
+// https://github.com/operator-framework/operator-registry/pull/86
+// is addressed, and remove entirely on a new release.
+replace github.com/operator-framework/operator-registry => github.com/estroz/operator-registry v1.3.1-0.20190912000639-78e83c9070d6
```
[the remark this answers is not in the document]
Comment (estroz) [reply go.mod]: Note: this has been merged.

**Comment by estroz** [comment]
@njhale @awgreene comments addressed, PTAL

File: go.mod [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+	github.com/sirupsen/logrus v1.4.2
+	github.com/spf13/cobra v0.0.5
+	k8s.io/apiextensions-apiserver v0.0.0-20190918161926-8f644eb6e783
+	k8s.io/apimachinery v0.0.0-20190913080033-27d36303b655
+	k8s.io/client-go v8.0.0+incompatible
+)
+
+// TODO(estroz): now that https://github.com/operator-framework/operator-registry/pull/86
+// has been addressed, remove this line once a new release is made.
+replace github.com/operator-framework/operator-registry => github.com/operator-framework/operator-registry v1.4.1-0.20191012191115-628ebd4ada3c
```
[the remark this answers is not in the document]
Comment (estroz) [reply go.mod]: Yep, updated this locally. Will push soon.

File: pkg/validation/internal/csv.go [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+
+type CSVValidator struct{}
+
+func (f CSVValidator) Validate(objs ...interface{}) (results []errors.ManifestResult) {
+	for _, obj := range objs {
+		switch v := obj.(type) {
+		case *operatorsv1alpha1.ClusterServiceVersion:
+			results = append(results, validateCSV(v))
+		case *registry.ClusterServiceVersion:
+			results = append(results, validateCSVRegistry(v))
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/csv.go]: I was thinking in a case like [this](https://github.com/operator-framework/operator-registry/blob/master/pkg/registry/bundle.go#L231). I'm fine with removing it though.

File: pkg/validation/internal/csv.go [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+// Returns updated error log with missing optional/mandatory field/struct objects.
+func updateLog(result *errors.ManifestResult, typeName string, newParentStructName string, emptyVal bool, isOptionalField bool) {
+	if !emptyVal {
+		return
+	}
+	if isOptionalField {
+		// TODO: update the value field (typeName).
+		result.Add(errors.WarnFieldMissing("", newParentStructName, typeName))
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/csv.go]: I think that's a reasonable addition.

File: pkg/validation/internal/csv.go [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+// checkFields runs checkMissingFields and returns its errors.
+func checkFields(csv *operatorsv1alpha1.ClusterServiceVersion) (errs []errors.Error) {
+	result := errors.ManifestResult{}
+	checkMissingFields(&result, reflect.ValueOf(csv), "")
+	return append(result.Errors, result.Warnings...)
+}
+
+// Recursive function that traverses a nested struct passed in as reflect value, and reports for errors/warnings
+// in case of null struct field values.
+func checkMissingFields(result *errors.ManifestResult, v reflect.Value, parentStructName string) {
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/csv.go]: Yeah this would be a nice internal function to use generally.

File: pkg/validation/interfaces/validator.go [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+	// one for each object validated.
+	Validate(...interface{}) []errors.ManifestResult
+}
+
+// Validators is a set of Validator's that can be run via Apply.
+type Validators []Validator
+
+// ApplyParallel invokes each Validator in vals, and collects and returns
+// the results.
+func (vals Validators) Apply(objs ...interface{}) (results []errors.ManifestResult) {
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/interfaces/validator.go]: We could, although keeping this as a non-`Validator` allows us to make changes to the signature, i.e. if we wanted to add functional options or a context.

File: README.md [hunk 23]
```diff
@@ -1,2 +1,19 @@
 # api
-Contains the API definitions used by OLM and Marketplace
+
+Contains the API definitions used by [Operator Lifecycle Manager][olm] (OLM) and [Marketplace Operator][marketplace]
+
+## `pkg/validation`: Operator Manifest Validation
+
+`pkg/validation` exposes a convenient set of interfaces to validate Kubernetes object manifests, primarily for use in an Operator project.
+
+Additionally this package defines a valid Operator [manifests format][registry-manifests] by exposing a set of `interfaces.Validator`'s to verify them and their constituent manifests. These bundles typically consist of a [Package Manifest][pkg-man] and a set of versioned [bundles][registry-manifests]. Each bundle contains a [ClusterServiceVersion][csv] and one or more [CustomResourceDefinition][crd]'s.
```
[the remark this answers is not in the document]
Comment (estroz) [reply README.md]: Done.

File: pkg/validation/internal/csv.go [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+
+		tag := fieldType.Tag.Get("json")
+		// Ignore fields that are subsets of a primitive field.
+		if tag == "" {
+			continue
+		}
+
+		// Omitted field tags will contain ",omitempty", and ignored tags will
+		// match "-" exactly, respectively.
+		isOptionalField := strings.Contains(tag, ",omitempty") || tag == "-"
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/csv.go]: Can you elaborate on what you mean by "OpenAPI spec for CRDs not included in the bundle" a bit? What manifest kind are we checking for missing fields in the context of your question?

File: pkg/validation/internal/csv.go [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+	"github.com/operator-framework/api/pkg/validation/errors"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+
+	operatorsv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+type CSVValidator struct{}
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/csv.go]: So something like

```go
type CSVValidator func(*operatorsv1alpha1.ClusterServiceVersion) errors.ManifestResult

func (f CSVValidator) Validate(objs ...interface{}) (results []errors.ManifestResult) {
	for _, obj := range objs {
		switch v := obj.(type) {
		case *operatorsv1alpha1.ClusterServiceVersion:
			results = append(results, f(v))
		}
	}
	return results
}
```

Then export functions that actually validate the CSV object?

File: README.md [hunk 26]
```diff
@@ -1,2 +1,10 @@
 # api
-Contains the API definitions used by OLM and Marketplace
+
+Contains the API definitions used by [Operator Lifecycle Manager][olm] (OLM) and [Marketplace Operator][marketplace]
+
+## `pkg/validation`: Operator Manifest Validation
+
+`pkg/validation` exposes a convenient set of interfaces to validate Kubernetes object manifests, primarily for use in an Operator project.
```
Comment (gallettilance) [inline README.md]: What do you think of adding some examples? This could be in a separate PR/folder

Comment (gallettilance) [reply README.md]: I'm thinking examples of implementing validators / using the apis

File: cmd/operator-verify/main.go [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/spf13/cobra"
+)
+
+func main() {
+	rootCmd := &cobra.Command{
+		Use:   "operator-verify",
+		Short: "Operator manifest validation tool",
+		Long:  `operator-verify is a CLI tool that calls functions in pkg/validation.`,
+	}
+
```
Comment (gallettilance) [inline cmd/operator-verify/main.go]: what do you think of adding an option to output the resulting errors/warnings to a log file - I think this was a use case that Ralph or Scott had with operator courier

File: pkg/internal/bundle.go [hunk 28]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+// AddOperatorBundle adds the package manifest to l.
+func (l *manifestsLoad) AddPackageChannels(pkg registry.PackageManifest) error {
+	l.pkg = pkg
+	return nil
+}
+
+// AddBundlePackageChannels is a no-op to implement the registry.Load interface.
+func (l *manifestsLoad) AddBundlePackageChannels(manifest registry.PackageManifest, bundle registry.Bundle) error {
```
Comment (gallettilance) [inline pkg/internal/bundle.go]: Note to self: As we change operator-registry we'll need to be aware of the dependency here on the implementation of its Load interface.

File: pkg/validation/validation.go [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/operator-framework/api/pkg/validation/internal"
+)
+
+// DefaultPackageManifestValidators returns a package manifest Validator that
+// can be run directly with Apply(objs...). Optionally, any additional
+// Validator's can be added to the returned Validators set.
+func DefaultPackageManifestValidators(vals ...interfaces.Validator) interfaces.Validators {
```
Comment (njhale) [inline pkg/validation/validation.go]: nit: This is a little confusing to me. I would expect this to be something like `WithDefaultPackageManifestValidator` or `AddDefaultPackageManifestValidator`. Though those don't really make sense w/ zero arguments, while `Default*` does.

File: README.md (hunk 26 above)
Comment (njhale) [reply README.md]: This might make a nice follow-up.

File: cmd/operator-verify/main.go (hunk 27 above)
Comment (njhale) [reply cmd/operator-verify/main.go]: I also think this can be a follow-up.

File: pkg/internal/bundle.go (hunk 28 above)
Comment (estroz) [reply pkg/internal/bundle.go]: In general we're going to have to do something about operator-registry types. If we want this to be a dependency of operator-registry, we're either going to have to copy over types or copy over parse code and use internal types. WDYT? We can probably clean this up in a follow-up PR.

File: pkg/validation/internal/manifests.go [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+			result.Add(errors.WarnInvalidCSV("`spec.replaces` field matches its own `metadata.name`. It should contain `metadata.name` of the old CSV to be replaced", csv.GetName()))
+		} else {
+			csvNames[csv.GetName()] = struct{}{}
+			replacesNames[csv.Spec.Replaces] = csv.GetName()
+		}
+	}
+	for replaces, sourceCSV := range replacesNames {
+		if _, csvExists := csvNames[replaces]; !csvExists {
+			result.Add(errors.ErrInvalidCSV(fmt.Sprintf("%q mentioned in the `spec.replaces` field is not present in manifests", replaces), sourceCSV))
+		}
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/manifests.go]: Ok I will consider `spec.skips` and `metadata.annotations["olm.skipRange"]` in this validator.

File: pkg/validation/internal/manifests.go [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+
+import (
+	"fmt"
+
+	"github.com/operator-framework/api/pkg/validation/errors"
+
+	"github.com/operator-framework/operator-registry/pkg/registry"
+)
+
+type ManifestsValidator struct{}
```
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/manifests.go]: I suppose this is a fairer name given this validator's current logic. However I'd rather keep the name general in case the constituent manifest types of a manifests directory are changed, or other, non-update graph validation logic is added in the future. How about `VersionedManifestSetValidator` or `ManifestsDirValidator`?

Alternatively we could add future logic as described above in another validator and reserve this one for validating the update graph.

File: pkg/validation/internal/csv.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/csv.go]: @njhale as a followup or in this PR? I'm thinking follow-up.

File: pkg/validation/validation.go (hunk 29 above)
Comment (estroz) [reply pkg/validation/validation.go]: I'm thinking we implement something like:

```go
type PackageManifestValidator struct{}

func (f PackageManifestValidator) Validate(objs ...interface{}) (results []errors.ManifestResult) {
    ...
}

func (f PackageManifestValidator) WithValidators(vals ...interfaces.Validator) interfaces.Validators {
	return append(vals, f)
}
```

Then we can do:

```go
var PackageManifestValidators = internal.PackageManifestValidator{}

func foo() {
	results := PackageManifestValidators.WithValidators(myvals...).Apply()
}
```

File: pkg/validation/internal/csv.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/validation/internal/csv.go]: Almost. What I'm proposing just allows functions to implement the `Validator` interface:

```go
type ValidatorFunc Validate(...interface{}) []errors.ManifestResult

func (v ValidatorFunc) Validate(objs ...interface{}) (results []errors.ManifestResult) {
    return v(objs...)
}

var CSVValidator Validator = ValidatorFunc(validateCSVs)

func validateCSVs(objs ...interface{}) (results []errors.ManifestResult) {
    for _, obj := range objs {
		switch v := obj.(type) {
		case *v1alpha1.ClusterServiceVersion:
			results = append(results, validateCSV(v))
		case *registry.ClusterServiceVersion:
			results = append(results, validateCSVRegistry(v))
		}
	}
	return results
}
```

File: pkg/validation/internal/csv.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/validation/internal/csv.go]: Let's leave it for now. We can always factor out later.

File: pkg/validation/validation.go (hunk 29 above)
Comment (njhale) [reply pkg/validation/validation.go]: I like it!

File: pkg/validation/internal/manifests.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/validation/internal/manifests.go]: > Alternatively we could add future logic as described above in another validator and reserve this one for validating the update graph.

This is more where my head is at, since I view validators as composable units that form larger validators (if `Validators` implemented `Validator`). I can also see where it can be preferable to make larger units to avoid duplication of work or handle interdependencies. If that's the case, then either of the names you've suggested are fine with me.

File: pkg/validation/internal/csv.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/validation/internal/csv.go]: One example would be the OpenAPI spec for a Deployment. We would validate the manifest against the OpenAPI schema, which would tell us when fields are missing (due to required directives, etc...). Does that make more sense?

[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/csv.go]: Yes that does. I will implement this in a follow-up.

File: pkg/validation/internal/manifests.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (estroz) [reply pkg/validation/internal/manifests.go]: Ehh because this code is only validating the update graph right now, we can name this `PackageUpdateGraphValidator` and add another validator later. These names are subject to change since they're internal types.
